### PR TITLE
Fix compatibility issue with incoming webhooks in Stripe API version 2022-11-15

### LINKED ISSUE #2488: Incompatibility with Stripe API version 2022-11-15
**Describe the bug**
When UPE is enabled and the latest version of Stripe's API (2022-11-15) is used, various issues and notices are caused, the severity of which depends on the payment method used.

**To Reproduce**
1. Set your Stripe account to use API version 2022-11-15
2. Enable the new checkout experience (UPE) and put Stripe in test mode.
3. Enabled multiple payment methods. I used CCs and a deferred method, Bancontact.
4. Place an order with:
* A 3DSv2 required CC
* A regular CC
* Bancontact
5. See varying errors depending on the payment method.

**3DSv2 Required Card**

This seemed to produce the most severe problems out of the 3 methods I tested with. The checkout flow doesn't present any errors. However, once the order is placed and you view it on the backend, no charge is linked and the order is set to On Hold. But, the payment in Stripe succeeded.

In the Stripe logs, this error was thrown: `Could not find order via charge ID: ch_3M9GazI21hhR58QI0NUM2UqF`

**Regular CC and Bancontact**

Both of these payment methods completed in Stripe and the order status was changed to processing. But multiple notices were thrown with WP_DEBUG enabled:

<details>

```
Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 294

Notice: Trying to get property 'id' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 294

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 298

Notice: Trying to get property 'id' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 298

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 304

Notice: Trying to get property 'balance_transaction' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 304

Notice: Trying to get property 'id' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 304

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 308

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 309
```

</details>

A different error was thrown in the Stripe logs: `Unable to update fees/net meta for order: 31`.

And the order itself had no charge linked at the top and the first charge ID in the order notes was empty:

A common symptom across different payment methods is an empty charge ID in the order notes, and this error in the order notes: `Stripe charge authorized (Charge ID: ). Process order to take payment, or cancel to remove the pre-authorization. Attempting to refund the order in part or in full will release the authorization and cancel the payment. Order status changed from Pending payment to On hold.`

**Expected behavior**
I expected UPE to be compatible with the latets verison of Stripe's API.

**Workaround**
~~Disabling UPE and/or switching to a Stripe API version before 2022-11-15 solves these issues.~~
Disabling UPE does not solve this issue. Stripe API versions can only be rolled back within 72 hours of upgrading.

**Environment (please complete the following information):**
 - WordPress Version: 6.1.0
 - WooCommerce Version: 7.1.0
 - Stripe Plugin Version: 7.0.1
 - Browser [e.g. chrome, safari] and Version
 - Any other plugins installed

**Additional context**
Identified in: 5741056-zen

SSR:
<details>

```
`
### WordPress Environment ###

WordPress address (URL): https://continental-galliform.jurassic.ninja
Site address (URL): https://continental-galliform.jurassic.ninja
WC Version: 7.1.0
REST API Version: ✔ 7.1.0
WC Blocks Version: ✔ 8.7.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.54 (Unix) OpenSSL/1.0.2g
PHP Version: 7.4.33
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.1.0
WC Database Prefix: wp_
Total Database Size: 5.80MB
Database Data Size: 4.22MB
Database Index Size: 1.58MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.17MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.08MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_bookings_availability: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_bookings_availabilitymeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_booking_relationships: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
bookable_person: 4
page: 7
post: 2
product: 5
revision: 1
shop_order: 11
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (5) ###

Bookings Helper: by WooCommerce – 1.0.3
Companion Plugin: by Osk – 1.28
Jetpack: by Automattic – 11.5.1
WooCommerce Stripe Gateway: by WooCommerce – 7.0.1
WooCommerce: by Automattic – 7.1.0

### Inactive Plugins (4) ###

Akismet Anti-Spam: by Automattic – 5.0.1
Hello Dolly: by Matt Mullenweg – 1.7.2
WooCommerce Bookings: by WooCommerce – 1.15.68
WooCommerce Google Analytics Integration: by WooCommerce – 1.5.17

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: booking (booking)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/
Checkout: #7 - /checkout/
My account: #8 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Three
Version: 1.0
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2022-11-29 18:06:00 +00:00
Options: ✔
Notes: 43
Onboarding: -

### Action Scheduler ###

Complete: 40
Oldest: 2022-11-25 18:17:56 +0000
Newest: 2022-11-28 23:59:21 +0000

Pending: 2
Oldest: 2022-11-29 22:53:37 +0000
Newest: 2022-12-25 23:12:57 +0000


### Status report information ###

Generated at: 2022-11-28 23:59:30 +00:00
`
```

</details>

### PR DESCRIPTION
**Pull request by ricardo**
Fixes #2488
More context: p1670013028525619/1668541074.915399-slack-CHG7MTCAF

**TL;DR:** Starting from [API version 2022-11-15](https://stripe.com/docs/upgrades#2022-11-15), the `charges` prop was removed from the PaymentIntent object.

In the plugin, the API version is [overwritten](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/1833465bc14d3a699abec9ec7d4f64b0fbdb4f24/includes/class-wc-stripe-api.php#L17) when making requests to the Stripe API, but for incoming webhooks, there's no way to overwrite the version. This PR makes the webhook handler compatible with the `charges` prop removal.

## Changes proposed in this Pull Request:

- If the `charges` prop is missing, make a request to the `charges` endpoint to expand the ID from `latest_charge`.

## Testing instructions

You'll need a standard Stripe account with its [API version](https://dashboard.stripe.com/developers) set to anything prior to [2022-11-15](https://stripe.com/docs/upgrades#2022-11-15) (not marked as default in the dashboard). The WCPay Dev (Express) account can be used if you don't have a personal test account.

1. Expose your local environment publicly. You can use Ngrok or JT. (For valet: `cd valet/stripe && valet unsecure && valet share`.
2. Create a new webhook endpoint from the [Stripe dashboard](https://dashboard.stripe.com/test/webhooks).
   i. Use the URL from the **Stripe plugin settings > Settings > Account details**.
   ii. **Add events > Select all events**.
   iii. Select API version 2022-11-15.
4. Add the endpoint and copy the webhook signing secret. Paste the secret into the plugin test webhook secret.
5. Enable UPE.
6. Test the checkout with Bancontact, a regular card and a 3DS card. Ensure the checkout doesn't return any errors, the charge succeeds in Stripe and there are no errors in the order notes from WC.

**Note:** There might be duplicate order notes for card payments processed from webhooks using the latest API version I'm investigating this as a separate issue.

If you test the above from `develop`, an error is thrown for Bancontact checkouts, and the order remains on-hold in WC.

You can make sure this is BW compatible with the previous versions by disabling the 2022-11-15 webhook and testing this PR with a different endpoint with an older version.

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

File: includes/class-wc-stripe-webhook-handler.php [hunk 1]
```diff
@@ -845,7 +845,8 @@ public function process_payment_intent_success( $notification ) {
 				break;
 			case 'payment_intent.succeeded':
 			case 'payment_intent.amount_capturable_updated':
-				$charge = end( $intent->charges->data );
```
Comment (cesarcosta99) [inline includes/class-wc-stripe-webhook-handler.php:848]: I understand we're fixing webhooks here, however there are lots of occurrences of `end( $intent->charges->data )` in the code. Should we replace those so that those pieces of code can benefit from this fix as well?

**Review by cesarcosta99 — APPROVED** [review]
LGTM!

Tested with latest API ✅ and tested it's backwards compatible ✅

>If you test the above from develop, an error is thrown for Bancontact checkouts, and the order remains on-hold in WC.

I could reproduce the errors, but it went out a bit different for me. The errors were thrown for all methods I tried (regular card, 3DS card and Bancontact). The order was also processed normally.

>There might be duplicate order notes for card payments processed from webhooks using the latest API version

I noticed that too, but it's also happening when using the previous API version. It looks a bit inconsistent, though. For some card payments the notes were not duplicated, and for some Bancontact payments they were.

I left a question regarding the approach used. Feel free to ignore if it's not relevant, or open a new issue/PR if it is.

File: includes/class-wc-stripe-webhook-handler.php (hunk 1 above)
Comment (ricardo) [reply includes/class-wc-stripe-webhook-handler.php:848]: This issue should only affect webhooks, because we're overwriting the API version [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/1833465bc14d3a699abec9ec7d4f64b0fbdb4f24/includes/class-wc-stripe-api.php#L17) when we make the requests ourselves.

When we update the plugin to the latest API, then we would want to add this and other changes.

**Comment by ricardo** [comment]
> The errors were thrown for all methods I tried (regular card, 3DS card and Bancontact). The order was also processed normally.

One thing I noticed while debugging the webhooks is that the errors or the "on-hold" status change might depend on how early the webhooks are received. If it's before the redirect to "order received", the results might change. There might be also some discrepancies for UPE vs non-UPE upon redirect, so I'd say this is expected.
